Introducing minishift - run openshift locally

budhrg opened the pull request:
Adding a blog post on [Minishift](https://github.com/minishift/minishift).

cc @LalatenduMohanty @thatdocslady @Preeticp

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+---
+
+It has been quite a few months since our [last blog post](http://www.projectatomic.io/blog/2016/08/vagrant-service-manager-1-3-0-release/) on [vagrant-service-manager](https://github.com/projectatomic/vagrant-service-manager) as we have been looking around the alternative of our [Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle). It would be a separate b ...
+
```
we have been looking around the alternative of our [Atomic Developer Bundle (ADB)]> we have been looking around for an alternative to [Atomic Developer Bundle (ADB)].
Better would be> we have been looking around for alternative solutions to replace the [Atomic Developer Bundle (ADB)].
A separate blog post will address the reasons for this.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+    ```
+
+    This will download the latest [_minishift-b2d-iso_](https://github.com/Minishift/minishift-b2d-iso), starts a VM using libmachine, download OpenShift client binary (`oc`) with version `1.4.1` and cache both `oc` and ISO into your `$HOME/.minishift/cache` folder and provisions OpenShift single node cluster.
+
````
This will: and then use bullet points for what it does, would make it easier to read.
starts a VM> start a VM, 
download the OpenShift client binary (`oc`), version `1.4.1`
and provision

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+    This will download the latest [_minishift-b2d-iso_](https://github.com/Minishift/minishift-b2d-iso), starts a VM using libmachine, download OpenShift client binary (`oc`) with version `1.4.1` and cache both `oc` and ISO into your `$HOME/.minishift/cache` folder and provisions OpenShift single node cluster.
+
+    This step also automatically logged you in with `developer` user in `myproject` project.
+
```
Suggest you add it to the above bullet point.
or 
This step also logs you in as a `developer` in the `myproject` project.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+    This step also automatically logged you in with `developer` user in `myproject` project.
+
+- Export the OpenShift client binary (`oc`) to _PATH_ environment variable as follows:
+
```
step 2
to the _PATH_....

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+    ```
+
+- Deploy a sample application as follows:
+
````
step 3
substeps for steps within this step.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+        $ oc expose svc/nodejs-ex
+        ```
+    - Open the application in browser as follows:
+
````
in a browser

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+        ```
+
+- When done with building application for the day, stop the cluster as follows:
+
````
step 4
building the...

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+    ```
+
+    Now, on some another day, if you want to access this cluster, just use `minishift start`.
+
````
If you want to access the same cluster some other day, use

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,110 @@
+    ```
+
+[Minishift's  README](https://github.com/Minishift/minishift#minishift) has more detailed description of getting start with single node OpenShift cluster in different OS platforms.
+
````
extra space before readme
getting started with a

budhrg replied:
Changed with this line "Refer [Minishift's README](https://github.com/Minishift/minishift#minishift) to get more detailed description of getting start with single node OpenShift cluster."

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+
+## Feedback
+
+Try Minishift! We would love to get your feedback. If you hit by any problem, you can open the issue in [minishift issue tracker](https://github.com/minishift/minishift/issues) but please search through it before creating it. It's possible that the similar issue is already open.
```
If you hit a problem, please raise an issue...
through the listed issues
a similar

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+Try Minishift! We would love to get your feedback. If you hit by any problem, you can open the issue in [minishift issue tracker](https://github.com/minishift/minishift/issues) but please search through it before creating it. It's possible that the similar issue is already open.
+
+Another way is to communicate with us through IRC as mentioned below.
+
```
you could also reach us at our IRC channel mentioned below.

budhrg commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+
+Minishift is actually a single binary which uses [libmachine](https://github.com/docker/machine/tree/master/libmachine) to provisions VM and [OpenShift Origin](https://github.com/openshift/origin) for running the cluster.
+
+It's a platform (OS) agnostic CLI tool and supports multiple hypervisors like VirtualBox, Xhyve (Mac OS), VMWare Fusion (Mac OS) and HyperV (Windows). Detailed instructions about it can be found in the [prerequisites section](https://github.com/Minishift/minishift#prerequisites) of [Minishift's README](https://github.com/minishift/minishift#minishift).
```
> add comma after agnostic.

`It's a platform (OS) agnostic CLI tool` is a single phrase.

> Why is KVM(fedora) not mentioned ?

This is generalized introduction of getting start. User can find more about OS specific in documentation. Hence didn't mentioned any OS.

budhrg commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+- Add the minishift binary to your PATH environment variable.
+
+Other alternative ways to install Minishift can be found [here](https://github.com/Minishift/minishift#installing-minishift).
+
```
@Preeticp  I think I need to add "To install in specific OS, refer to [minishift the installation guide](https://github.com/Minishift/minishift#installing-minishift)."

budhrg commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,110 @@
+- Download the archive for your OS from the Minishift [releases page](https://github.com/minishift/minishift/releases) and unpack it.
+- Copy the contents of the directory to your preferred location.
+- Add the minishift binary to your PATH environment variable.
+
```
Done

thatdocslady replied:
Looks like this is still an unordered list, @budhrg

budhrg replied:
Done

thatdocslady commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+- Copy the contents of the directory to your preferred location.
+- Add the minishift binary to your PATH environment variable.
+
+To install in specific OS, refer to the [minishift installation guide](https://github.com/Minishift/minishift#installing-minishift).
```
Please note PR https://github.com/minishift/minishift/pull/621, this link (and others) will need to change after we merge.

budhrg replied:
@thatdocslady We can update that part later if this PR is getting merge soon 😄

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+tags: minishift, atomic-developer-bundle, openshift-local
+---
+
+It has been quite a few months since our [last blog post](http://www.projectatomic.io/blog/2016/08/vagrant-service-manager-1-3-0-release/) on [vagrant-service-manager](https://github.com/projectatomic/vagrant-service-manager) as we have been looking around for alternative solutions to replace the [Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle). A sepa ...
```
I do not think we need to mention about our last blog post on VSM. We should communicate that we have came  up with a better user experience in Minishift as compared to ADB. Also the sentence about "have been looking around for alternative solutions to replace the ADB" is wrong. We wanted to improve the user experience and  also wanted to address the issues users were facing with ADB.

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+
+It has been quite a few months since our [last blog post](http://www.projectatomic.io/blog/2016/08/vagrant-service-manager-1-3-0-release/) on [vagrant-service-manager](https://github.com/projectatomic/vagrant-service-manager) as we have been looking around for alternative solutions to replace the [Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle). A sepa ...
+
+With this blog post, we introduce [Minishift](https://github.com/minishift/minishift), an alternative to ADB, which is a tool that helps you run OpenShift locally by running a single-node OpenShift cluster inside a VM. You can try out OpenShift or develop with it, day-to-day, on your local host.
```
We should remove `an alternative to ADB` as we mentioned above that we moved to Minishift as it provides better user experience.

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+With this blog post, we introduce [Minishift](https://github.com/minishift/minishift), an alternative to ADB, which is a tool that helps you run OpenShift locally by running a single-node OpenShift cluster inside a VM. You can try out OpenShift or develop with it, day-to-day, on your local host.
+
+Minishift is actually a single binary which uses [libmachine](https://github.com/docker/machine/tree/master/libmachine) to provision a VM with a local, single node, [OpenShift Origin](https://github.com/openshift/origin) cluster.
+
```
We should mention about `oc cluster up` here i.e. we use `cluster up` to provision the OpenShift inside the VM. If we can explain it little more it will be good.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+## Quickstart with Minishift
+
+1. Get start as follows:
+
```
Get started

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
````diff
@@ -0,0 +1,116 @@
+    ```
+
+Refer [Minishift's README](https://github.com/Minishift/minishift#minishift) to get more detailed description of getting start with single node OpenShift cluster.
+
````
getting started with a

thatdocslady replied:
"Refer to the [Minishift README]" (also, pinning this link TBC when we publish docs)

jberkus commented:
Let me know when the edits are done, thanks!

budhrg commented:
Hi @LalatenduMohanty , address your suggestions. Also, beginning with different note.

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,120 @@
+
+[Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle) has been a great tool so far in providing a prepackaged development environment filled with production-grade, pre-configured tools, that makes the container developer's lives easier.
+
+However, many of our users experienced that the downloads of the vagrant box `projectatomic/adb` of around 500 MB sometimes take time and even fails to download in a slow networks. This and few other potential issues which I would like to address in my another blog post impelled us to think about the better user experience.
```
@budhrg The download size was not the biggest issue with ADB, but the issues which are Vagrant related. I do not see any value this text adds to the overall blog post. We should just tell Minishift is a better tool than ADB  because of overall user experience. I will not suggest you to elaborate on this in this blog

budhrg replied:
Ya @LalatenduMohanty , doesn't make sense to give half-baked reasons 👍

Preeticp replied:
Suggest replacing line 12 and 14 with :
We would now like to introduce you to Minishift(link) which provides a far better user experience as compared to ADB. We moved from ADB to Minishift, to improve the user experience as well as to address Vagrant based issues being faced by our users.

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,120 @@
+However, many of our users experienced that the downloads of the vagrant box `projectatomic/adb` of around 500 MB sometimes take time and even fails to download in a slow networks. This and few other potential issues which I would like to address in my another blog post impelled us to think about the better user experience.
+
+From the past few months, we have been looking and investigating for better solutions which could give user a better experience.
+
```
IMO this is also extra detail, which an user will not be interested much. Rather we should mention that Minishift is a drastic improvement on ADB and we have been working on Minishift from last few months.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,120 @@
+
+From the past few months, we have been looking and investigating for better solutions which could give user a better experience.
+
+With this blog post, we introduce [Minishift](https://github.com/minishift/minishift), which is a CLI tool that helps you run OpenShift locally by running a single-node cluster inside a VM. You can try out OpenShift or develop with it, day-to-day, on your local host.
```
If you use my suggestion above, then replace 'With this blog post, we introduce [Minishift](https://github.com/minishift/minishift), which' with 'Minishift is a CLI....'

budhrg commented:
@LalatenduMohanty Update based on @Preeticp 's suggestion.

Preeticp commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,118 @@
+
+[Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle) has been a great tool so far in providing a prepackaged development environment filled with production-grade, pre-configured tools, that makes the container developer's lives easier.
+
+We would now like to introduce you to Minishift](https://github.com/minishift/minishift) which provides a far better user experience as compared to ADB. We moved from ADB to Minishift, to improve the user experience as well as to address Vagrant based issues being faced by our users.
```
[Minishift]

budhrg replied:
Oops my bad :)

jberkus commented:
So, is this ready?

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,118 @@
+---
+
+[Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle) has been a great tool so far in providing a prepackaged development environment filled with production-grade, pre-configured tools, that makes the container developer's lives easier.
+
```
The above paragraph is not required IMO. Also there is a disconnect between the paragraphs i.e. above and below. The below paragraph looks good to start the blog.

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,118 @@
+[Atomic Developer Bundle (ADB)](https://github.com/projectatomic/adb-atomic-developer-bundle) has been a great tool so far in providing a prepackaged development environment filled with production-grade, pre-configured tools, that makes the container developer's lives easier.
+
+We would now like to introduce you to [Minishift](https://github.com/minishift/minishift) which provides a far better user experience as compared to ADB. We moved from ADB to Minishift, to improve the user experience as well as to address Vagrant based issues being faced by our users.
+
```
s/based/related/

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,118 @@
+Minishift is a CLI tool that helps you run OpenShift locally by running a single-node cluster inside a VM. You can try out OpenShift or develop with it, day-to-day, on your local host.
+
+It is actually a single binary which uses [libmachine](https://github.com/docker/machine/tree/master/libmachine) to provision a VM with a local, single node, [OpenShift Origin](https://github.com/openshift/origin) cluster. It uses [OpenShift local cluster management](https://github.com/openshift/origin/blob/master/docs/cluster_up_down.md) underneath to bring up single node cluster.
+
```
👍

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,118 @@
+
+It is actually a single binary which uses [libmachine](https://github.com/docker/machine/tree/master/libmachine) to provision a VM with a local, single node, [OpenShift Origin](https://github.com/openshift/origin) cluster. It uses [OpenShift local cluster management](https://github.com/openshift/origin/blob/master/docs/cluster_up_down.md) underneath to bring up single node cluster.
+
+It's a platform agnostic CLI tool that supports multiple hypervisors like KVM for Linux, Xhyve and VMWare Fusion for Mac OS, HyperV for Windows and VirtualBox in all OS platforms. Detailed instructions about it can be found in the [Minishift documentation](https://github.com/Minishift/minishift#prerequisites).
```
We have not tested VMWare Fusion, so lets not mention this.

budhrg replied:
I have taken that from our README @LalatenduMohanty , let's remove from that too.

budhrg commented:
@jberkus Seems like good to go now. WDYT @LalatenduMohanty  ?

LalatenduMohanty commented on source/blog/2017-03-30-minishift-intro.html.md:
```diff
@@ -0,0 +1,116 @@
+## Resources
+- Github Repo: https://github.com/minishift
+- Documentation: https://github.com/minishift/minishift#documentation
+- OpenShift Origin: https://www.openshift.org/minishift
```
This link description is wrong. It is not the link for OpenShift Origin. Sorry for not noticing in previous reviews.

LalatenduMohanty replied:
I would suggest we remove the link or rename it as "Minishift Home Page"

budhrg replied:
+1 for Minishift Home Page

budhrg replied:
Done @LalatenduMohanty

budhrg commented:
@jberkus , its ready to merge now 👍

jberkus commented:
Question: the article says that we're replacing ADB with Minishift.  However, it does not address the other container tools which used to be available with ADB in the various editions (e.g. Mesos edition).  Can you include a comment on that?  For example, will ADB continue to be available or stop being built?

jberkus commented:
If you can't answer those questions (about ADB), then it's probably better not to discuss replacing ADB with Minishift in the post.

budhrg commented:
@jberkus 

> If you can't answer those questions (about ADB), then it's probably better not to discuss replacing ADB with Minishift in the post.

We had added one line of text at the moment as : 
```
We moved from ADB to Minishift, to improve the user experience as well as to address Vagrant related issues being faced by our users.
```

I am planning to add a full blog post to address the reasons for this move. We discussed and agreed not to add much details about ADB in this blog post since this blog post is mostly about minishift introduction. 

Would it be fine?

jberkus commented:
How about: "We've shifted development effort from ADB to Minishift, to improve the user experience as well as to address Vagrant related issues being faced by our users.  We'll explain this more in a later blog post."

budhrg commented:
> How about: "We've shifted development effort from ADB to Minishift, to improve the user experience as well as to address Vagrant related issues being faced by our users. We'll explain this more in a later blog post."

Perfect texts. Thanks @jberkus :)

jberkus commented:
I've started copy-editing this.  However, I want to test out the instructions before I publish, so look for it tommorrow after I have a chance to test.

budhrg commented:
> I've started copy-editing this. However, I want to test out the instructions before I publish, so look for it tommorrow after I have a chance to test.

Sure 👍

jberkus commented:
OK, some technical issues with the instructions:

1. the version of `oc` is a problem, in that your instructions are version-specific, but the download give me the latest version of the tools, whatever they are.  For example, I ran the instructions today, which resulted in me getting the 1.5.0-rc.0 version of `oc`, causing the PATH command to fail.

```
[josh@redhat bin]$ export PATH=$PATH:$HOME/.minishift/cache/oc/v1.4.1
[josh@redhat bin]$ oc new-app https://github.com/openshift/nodejs-ex -l name=myapp
bash: oc: command not found...
[josh@redhat bin]$ ls ~/.minishift/cache/oc/v1.5.0-rc.0/oc 
/home/josh/.minishift/cache/oc/v1.5.0-rc.0/oc
```

Why doesn't minishift setup just set the path for `oc`?  It sets the path for everything else.  Regardless, we're going to need to alter the step-by-step instructions to account for the user having a different version number.

2. I got the following error on minishift deployment:

```
E0407 11:53:31.520005    9621 start.go:205] Error creating Minishift kubeconfig Error reading config file /home/josh/.minishift/machines/minishift_kubeconfig
```

As far as I can tell, this error doesn't seem to have actually broken anything.

3. The `minishift service` command doesn't work, and fails with a misspelled error message:

```
[josh@redhat bin]$ minishift openshift service nodejs-ex -n myproject
Namespace myproject doesn't exits
[josh@redhat bin]$ minishift openshift service nodejs-ex
Namespace default doesn't exits
```
This may or may not be related to error (2)

Please comment on the above errors.  At a minimum, we need to deal with (1); (2) and (3) may be just my laptop.

jberkus commented:
@budhrg ?

budhrg commented:
@jberkus  Sorry for late reply

> The version of oc is a problem, in that your instructions are version-specific, but the download give me the latest version of the tools, whatever they are. For example, I ran the instructions today, which resulted in me getting the 1.5.0-rc.0 version of oc, causing the PATH command to fail.

Yes you are correct. We have updated with minimum 1.5.0-rc.0 version of openshift since it contain more out of box support for various feature like proxy etc. And, hoping it will remain quite some time. Also, added the text in bracket to inform user that this version might change based on minishift version. Hope it will help user.

> I got the following error on minishift deployment:

This seems to be setup problem.

> The minishift service command doesn't work, and fails with a misspelled error message:

Yes it is related to above error as we are reading `minishift_kubeconfig` to get admin details and running `oc get route` here for serving application url.

I have updated the PR to accommodate changes for version. Let me know if you need any further changes 

Thanks

jberkus commented:
Hmmm.  So the blog still needs to cover for future changes of version numbers.  I'll edit the text.

budhrg commented:
@jberkus , any update on it?

jberkus commented:
At this point, it'll go out Thursday US time.  I'm going to need to do some editing because the version-specific changes aren't clear.

Some of the stuff I asked about amount to usability issues with Minishift itself.  Where can I file bugs on these?

budhrg commented:
Hi @jberkus , 

You can file issue here https://github.com/minishift/minishift/issues . Update blog for same info too.

jberkus commented:
Ah, looks like the issue I wanted to file is already slated to be fixed in 1.1. Yay!

budhrg commented:
> Ah, looks like the issue I wanted to file is already slated to be fixed in 1.1. Yay!

Good to hear that. Which one BTW ? :smile:

budhrg commented:
@jberkus  Please hold on this PR we are about to release our 1st major release in next week around. 

I think it would be make sense to release the blog that time.

Please feel free to close this PR.

jberkus commented:
Won't we be using the same blog text, more-or-less?

budhrg commented:
Closing it now in favor of new PR
